### DISCUSSION TITLE
Loop over an Empty File Does Not Produce an Error

### DISCUSSION THREAD

**Discussion by grkal** [groops-devs/groops#233] [Q&A] [answered]
Dear GROOPS Team,

The topic of this message concerns the performance of the “LoopPrograms” program over an empty file.

In the “Loop” class’ description of the GROOPS documentation, it is mentioned that when the “FileAscii” config element is selected the loop takes place over a list of strings included in the specified file. Alternatively, if the “ManualList” is selected, then the loop takes place over a list of strings which are specified by the user.

A really nice feature of GROOPS which I have been using is to place the “LoopPrograms” program with the “ManualList” config element selected inside the “GroupPrograms” program where for the latter the “catchErrors” config element is specified as “< enabled >”. This way, I can overcome the potential problem of an empty manual list and continue with another loop/program specified in the “errorProgram”.

I thought that the same performance would be achieved when using the “LoopPrograms” program with the “FileAscii” config element selected inside the “GroupPrograms” program. However, when an empty file is used, the “errorProgram” is not executed which in my opinion should be the case since the file does not contain any string. At the same time and as it is expected, the loop does not execute the program which is specified inside it.

Attached you will find the “LoopOverEmptyFile.zip” file which includes the empty file “EmptyFile.txt” and the “LoopOverEmptyFile.xml” file which is a simple script I created so as to demonstrate the different performance when the loop takes place over an empty file versus an empty manual list.

As you will see, for the case of the loop over the empty file neither the program specified inside the loop nor the program specified as “errorProgram” is executed; however, for the case of the loop over the empty manual list the “errorProgram” program is executed and the following message appears:

"_--- GroupPrograms (Check with an empty manual list) ---
in D:\GROOPS\source\config\config.cpp:739 (run)
in D:\GROOPS\source\programs\system\loopPrograms.cpp:175 (run)
in D:\GROOPS\source\classes\loop\loop.cpp:53 (readConfig<std::shared_ptr<Loop> >)
in D:\GROOPS\source\classes\loop\loop.cpp:101 (create)
in D:\GROOPS\source\classes\loop\loopManualList.h:66 (LoopManualList)
in D:\GROOPS\source\config\config.cpp:819 (hasName)
config 'groops.program(GroupPrograms).program(LoopPrograms).loop(manualList)' must contain 'string'  continue with error programs..._"

Is there any edit that should take place on the “_source\config\config.cpp_” and/or the “_source\classes\loop\loopFileAscii.h_” files so as for the loop to produce an error when an empty file is used or is there something which I have not understood correctly?

It would be really useful if an error was triggered when an empty file is used.
 
Thank you very much in advance for your time and looking forward to your feedback.

Best regards,
Grigorios Kalimeris
Chair of Astronomical and Physical Geodesy
Technical University of Munich

[LoopOverEmptyFile.zip](https://github.com/groops-devs/groops/files/13648562/LoopOverEmptyFile.zip)

**Comment by tmayerguerr** [comment, marked as answer]
Hi Grigorios,

Sorry for my late answer.

I think groops does what I would expect: I know of no programming language in which an empty loop generates an error. Creating an empty ManualList, on the other hand, is pointless because the result is known at startup. This should not actually be configurable in the GUI.

If you want to change the source code you can add the line
```
  if(!count()) throw(Exception("empty loop"));
```
in `classes/loop/loopFileAscii.h` at the end of the constructor (near line 93).

Alternatively, the behavior can also be imitated in the script:
Create a [loop->ManualList](https://groops-devs.github.io/groops/html/loopType.html#manualList) and a [loop->fileAscii](https://groops-devs.github.io/groops/html/loopType.html#fileAscii) in the global section.
Then loop the element `string` with loopfileAscii (set loop) and set the value `string={loopString}`.

regards
Torsten

**Reply by grkal** [reply]
Dear Torsten,

Thank you very much for your detailed reply and for the nice solutions you provide.

I agree with you that creating an empty "ManualList" is pointless. However, my thought process behind my question is a case where the user sets up a “LoopPrograms” program with the “ManualList” config element selected with each "string" config element of the manual list being activated only when a user-specified condition is true (using the "Set Condition ..." feature).

If all the user-specified conditions which are connected to each "string" config element are false, then this would lead to an "empty" manual list. The error message which would be generated in this case helps because the user might either want the program to stop if no user-specified condition is true or to run the “errorProgram” if the loop is located inside the “GroupPrograms” program where the “catchErrors” config element is specified as “< enabled >”

Similarly, for the “LoopPrograms” program with the “FileAscii” config element selected, the user might have set in a previous part of the script a "RunCommand" program which assigns values to the file which will be called upon the loop, only if certain user-specified conditions are true. Then, if the user would again want the program to stop if the file is empty or to run another program, an error generated would help in this case. However, this is now possible with the workaround you provided which combines the [loop->ManualList](https://groops-devs.github.io/groops/html/loopType.html#manualList) and the [loop->fileAscii](https://groops-devs.github.io/groops/html/loopType.html#fileAscii) loops.

Best regards,
Greg
